builtin/providers/aws/resource_aws_opsworks_instance.go

```diff
@@ -0,0 +1,998 @@
+						},
+
+						"volume_type": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+							Computed: true,
+							ForceNew: true,
+						},
+					},
+				},
+				Set: func(v interface{}) int {
+					// there can be only one root device; no need to hash anything
+					return 0
+				},
+			},
+		},
+	}
+}
+
+func validateArchitecture(v interface{}, k string) (ws []string, errors []error) {
+	value := v.(string)
+	if value != "x86_64" && value != "i386" {
+		errors = append(errors, fmt.Errorf(
+			"%q must be one of \"x86_64\" or \"i386\"", k))
+	}
+	return
+}
+
+func validateAutoScalingType(v interface{}, k string) (ws []string, errors []error) {
+	value := v.(string)
+	if value != "load" && value != "timer" {
+		errors = append(errors, fmt.Errorf(
+			"%q must be one of \"load\" or \"timer\"", k))
+	}
+	return
+}
+
+func validateRootDeviceType(v interface{}, k string) (ws []string, errors []error) {
+	value := v.(string)
+	if value != "ebs" && value != "instance-store" {
+		errors = append(errors, fmt.Errorf(
+			"%q must be one of \"ebs\" or \"instance-store\"", k))
+	}
+	return
+}
+
+func validateState(v interface{}, k string) (ws []string, errors []error) {
+	value := v.(string)
+	if value != "running" && value != "stopped" {
+		errors = append(errors, fmt.Errorf(
+			"%q must be one of \"running\" or \"stopped\"", k))
+	}
+	return
+}
+
+func validateVirtualizationType(v interface{}, k string) (ws []string, errors []error) {
+	value := v.(string)
+	if value != "paravirtual" && value != "hvm" {
+		errors = append(errors, fmt.Errorf(
+			"%q must be one of \"paravirtual\" or \"hvm\"", k))
+	}
+	return
+}
+
+func resourceAwsOpsworksInstanceValidate(d *schema.ResourceData) error {
+	if d.HasChange("ami_id") {
+		if v, ok := d.GetOk("os"); ok {
+			if v.(string) != "Custom" {
+				return fmt.Errorf("OS must be \"Custom\" when using using a custom ami_id")
+			}
+		}
+
+		if _, ok := d.GetOk("root_block_device"); ok {
+			return fmt.Errorf("Cannot specify root_block_device when using a custom ami_id.")
+		}
+
+		if _, ok := d.GetOk("ebs_block_device"); ok {
+			return fmt.Errorf("Cannot specify ebs_block_device when using a custom ami_id.")
+		}
+
+		if _, ok := d.GetOk("ephemeral_block_device"); ok {
+			return fmt.Errorf("Cannot specify ephemeral_block_device when using a custom ami_id.")
+		}
+	}
+	return nil
+}
+
+func resourceAwsOpsworksInstanceRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*AWSClient).opsworksconn
+
+	req := &opsworks.DescribeInstancesInput{
+		InstanceIds: []*string{
+			aws.String(d.Id()),
+		},
+	}
+
+	log.Printf("[DEBUG] Reading OpsWorks instance: %s", d.Id())
+
+	resp, err := client.DescribeInstances(req)
+	if err != nil {
+		if awserr, ok := err.(awserr.Error); ok {
+			if awserr.Code() == "ResourceNotFoundException" {
+				d.SetId("")
+				return nil
+			}
+		}
+		return err
+	}
+
+	instance := resp.Instances[0]
+	instanceId := *instance.InstanceId
```

Comment:
We've learned in our time with the AWS API that all sorts of different shapes of responses are possible! :grinning: 

So we tend to be extra defensive in handling types out of the SDK - I'd double check both the length of `resp.Instances` and that the `InstanceId` pointer is non-null. (For the rest of the `d.Set` calls below the null check is handled in the implementation so we're good there.)